src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/AssemblyKernelExtension.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.Interactive;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.CSharp;
+using Microsoft.DotNet.Interactive.Utility;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Sql;
+using Microsoft.Spark.Utils;
+
+namespace Microsoft.Spark.Extensions.DotNet.Interactive
+{
+    /// <summary>
+    /// A kernel extension when using .NET for Apache Spark with Microsoft.DotNet.Interactive
+    /// Adds nuget and assembly dependencies to the default <see cref="SparkSession"/>
+    /// using <see cref="SparkContext.AddFile(string, bool)"/>.
+    /// </summary>
+    public class AssemblyKernelExtension : IKernelExtension
+    {
+        private const string TempDirEnvVar = "DOTNET_SPARK_EXTENSION_INTERACTIVE_TMPDIR";
+
+        private readonly PackageHelper _packageHelper = new PackageHelper(new PackageResolver());
+
+        /// <summary>
+        /// Called by the Microsoft.DotNet.Interactive Assembly Extension Loader.
+        /// </summary>
+        /// <param name="kernel">The kernel calling this method.</param>
+        /// <returns><see cref="Task.CompletedTask"/> when extension is loaded.</returns>
+        public Task OnLoadAsync(IKernel kernel)
+        {
+            if (kernel is CompositeKernel kernelBase)
+            {
+                Environment.SetEnvironmentVariable(Constants.RunningREPLEnvVar, "true");
+
+                DirectoryInfo tempDir = CreateTempDirectory();
+                kernelBase.RegisterForDisposal(new DisposableDirectory(tempDir));
+
+                kernelBase.AddMiddleware(async (command, context, next) =>
+                {
+                    if ((context.HandlingKernel is CSharpKernel kernel) &&
+                        (command is SubmitCode) &&
+                        TryGetSparkSession(out SparkSession sparkSession))
+                    {
```

Comment:
nit: should we factor out this block to a function?

Reply:
which block are you referring to ? Emitting the compilation dlls or the `PackageHelper` calls, or both ?

Reply:
I was thinking starting from `Compilation preCompilation = kernel.ScriptState.Script.GetCompilation();`. WDYT?

Reply:
or the whole middle ware code is also fine if it makes more sense.

Reply:
Moved emitting dll to a separate method.